### DISCUSSION TITLE
Recreating the Waffle Mark with the Dot mark

### DISCUSSION THREAD

**Discussion by jamesking** [observablehq/plot#2167] [Q&A] [answered]
I have a dataset that I would like to represent in a waffle-mark like way, but I would like each dot (or waffle "cell") to represent an individual datapoint, rather than a unit of a quantity. 

I think the solution would be to combine the dot mark with transforms or offsets to stack the dots in rows and columns to mimic the way that the waffle mark creates a grid.

The motivation is so the user can interact with each datapoint. In the screenshot below, the user is only able to interact with the area of the waffle mark that represents each _group_ of datapoints and not the individual datapoints themselves. 

![CleanShot 2024-09-10 at 14 27 18@2x](https://github.com/user-attachments/assets/9dcbccc0-2224-408a-80af-adbf7a5d0dc1)

I can see how I could use a `stack` transform to create a stack of dots, but I can't see how I could get to a waffle-like layout of individual dot marks. 

Any suggestions greatly appreciated.

**Comment by jamesking** [comment]
I have made some progress. The following gives me an arrangement of dot marks in a waffle-like grid, `50` "cells" wide. This is more-or-less what I want. The only problem is that I would like the Y-axis to have appropriate tick mark values which would the current value multiplied by `50`. 

```javascript
Plot.plot({
  aspectRatio: 1,
  marginLeft: 30,
  inset: 10,
  x: { axis: null },
  y: {},
  marks: [
    Plot.dot(
      data,
      Plot.stackY2({
        x: (d) => data.indexOf(d) % 50,
        y: (d) => 1,
        fill: "#ccc",
        r: 5
      })
    )
  ]
})
```
![CleanShot 2024-09-10 at 15 41 12@2x](https://github.com/user-attachments/assets/94845984-0606-4c00-a622-6175cccd6071)

**Reply by Fil** [reply]
maybe with `y: (d) => 1 / 50`?

The alternative would be to fix the waffle mark’s **tip** option, see #2132 ?

**Reply by jamesking** [reply, marked as answer]
@Fil thanks

> maybe with `y: (d) => 1 / 50`?

I have currently solved my tick mark issue with `tickFormat: (t) => (t - 1) * 50`. Your suggestion could also work, but I would also need to also adjust the `aspectRatio` to `1/50`. 

![CleanShot 2024-09-10 at 16 17 30@2x](https://github.com/user-attachments/assets/40a7b0e2-66b6-4c9d-879d-406e779ea282)

**Reply by jamesking** [reply]
@Fil from reading #2132 the `tip` option seems to apply to the waffle mark "area" whereas I want to address each individual dot directly: 

![CleanShot 2024-09-10 at 16 21 47@2x](https://github.com/user-attachments/assets/38a53bb0-a2ec-4c69-a2e5-762dc8331383)

**Reply by Fil** [reply]
Sure, in that case you wouldn't group and instead create 1 cell for each element.

**Reply by jamesking** [reply]
I don't think I would be able to colour groups of datapoints in that case (as in the above screenshot). Am I missing something?

**Reply by Fil** [reply]
You can color individual points; the difference with 👆 is that you won't group them. There is an example of this in the unit tests of #2132.

**Reply by jamesking** [reply]
That does look really good. Looking forward to trying out the updated waffle mark once #2132 lands.

**Reply by fcontrepois** [reply]
> Sure, in that case you wouldn't group and instead create 1 cell for each element.

What would the Plot code look like. I am trying to create a "bar chart" of waffles and would love to be have the tip showing the name of the service that the square represents. 

<img width="1205" height="560" alt="image" src="https://github.com/user-attachments/assets/3035c853-448a-40db-aa88-f79a9a2abed7" />


```
Plot.plot({
  width,
  x: {label: null, tickRotate: -45},
  y: { label: "# Services" },
  marginBottom: 80,
  marks: [
    Plot.waffleY(
     region_to_service,
     Plot.groupX({
      y: "count",
     }, 
     {
      x: "regioncode",
      y: "service",
      sort: {x: "-y"},
      fill: "var(--theme-foreground-focus)",
      tip: true,
     })),
  ]
})
```

**Reply by Fil** [reply]
@fcontrepois probably remove the `Plot.groupX` transform?

**Reply by fcontrepois** [reply]
Hello @Fil , and thanks for rapid answer. Should I move this to the forum section?

You can find the implementation in https://github.com/fcontrepois/aws.frankcontrepois.com

I changed the code to:
```js
Plot.plot({
  width,
  x: {label: null, tickRotate: -45},
  y: { label: "# Services" },
  marginBottom: 80,
  marks: [
    Plot.waffleY(
     region_to_service,
     {
      x: "regioncode",
      y: "service",
      sort: {x: "-y"},
      fill: "var(--theme-foreground-focus)",
      tip: true,
     }),
  ]
})
```

and the resulting graph is empty
<img width="1159" height="528" alt="image" src="https://github.com/user-attachments/assets/9715860f-455b-4271-b1f8-bfaf70588f51" />

the data is like:
```
Array(8463) [
  0: Object {
  regioncode: "ap-northeast-1"
  service: "acm-pca"
}
  1: Object {
  regioncode: "ap-northeast-1"
  service: "chime"
}
```

**Reply by Fil** [reply]
Can you try `y: 1`? you need to indicate the quantity for each square

**Reply by fcontrepois** [reply]
< wrong>
I think that is my issue. Service is not a number but a string, hence the groupping. What I was trying to get to is to have a tip showing the name of the service.

It seems that what the Waffle mark does is to take the number and create as many elements as needed but losing the connection to the data the individual square is representing.
< /wrong>

**Reply by fcontrepois** [reply]
Surprisingly the bars are still there...I am very confused. I will do a clean and retry

```js
Plot.plot({
  width,
  x: {label: null, tickRotate: -45},
  y: { label: "# Services" },
  marginBottom: 80,
  marks: [
    Plot.waffleY(
     region_to_service,
     {
      x: "regioncode",
      y: 1,
      sort: {x: "-y"},
      fill: "var(--theme-foreground-focus)",
      tip: true,
     }),
  ]
})
```

<img width="1147" height="445" alt="image" src="https://github.com/user-attachments/assets/adc909ae-be37-4709-ab3b-2ae5d75f0806" />

**Reply by Fil** [reply]
You can add the service name in the `fill` channel, or in an external channel if you only want to display it in the interactive tip ; and you can suppress the uninformative value of `y` in the tip. See https://observablehq.com/@recifs/plot-waffle-with-individual-points
